### PR TITLE
1083 973 date picker pop issues fixed

### LINKED ISSUE #1083: Datacapture, date picker doesn't open after cancel date picker dialog on same field
**Describe the bug**
When clicking on the date field like birthday, the date picker opens but when canceling dialog and clicking again then didn't open the date picker.  

**Component**
SDC library, Demo app

**To Reproduce**
1. Open Demo app.
2. Click on the plus icon.
3. Click on Date of Birth.
4. Click on the cancel button of the dialog.
5. Again click on Date of Birth.

**Expected behavior**
Open date picker dialog.

**Screenshots**
https://user-images.githubusercontent.com/98520274/151371327-b704bb6b-7053-460a-9af5-cfd54e7cef24.mp4

**Smartphone (please complete the following information):**
 - Device: Moto G(9)
 - Android version: 11
 - Build number: RPXS31.Q2-58-17-4-1
 - Version 1.0

**Additional context**

**Would you like to work on the issue?**
Please state if this issue should be assigned to you or who you think could help to solve this issue.

### PR DESCRIPTION
**Pull request by PallaviGanorkar**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #1083 #973

**Description**
code changes covers below ,
1. Date and time picker widget will be opened only on click of icon on right side in date and time input widget.
2. user is able to enter date and time using keyboard as well on click of input widget.
3. For date input widget, helper text is show from json if available otherwise show date format based on device locale as helper text

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: Bug fix 

**Screenshots (if applicable)**

**Checklist**
- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/docs/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://github.com/google/android-fhir/wiki/Contributing) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
@santosh-pingle can you take a look at this first? thanks

**Comment by santosh-pingle** [comment]
looking into it.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt [hunk 1]
```diff
@@ -130,6 +133,17 @@ internal val Questionnaire.QuestionnaireItemComponent.isHidden: Boolean
     return false
   }
 
+/** Whether the QuestionnaireItem should have entry format string. */
+val Questionnaire.QuestionnaireItemComponent.dateEntryFormat: String?
```
Comment (santosh-pingle) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:137]: Can we please write tests for it?

File: datacapture/src/main/java/com/google/android/fhir/datacapture/utilities/MoreLocalDate.kt [hunk 2]
```diff
@@ -33,3 +36,18 @@ internal val LocalDate.localizedString: String
...
+/** Gives date format for current locale of device */
+fun getDefaultDatePattern(): String {
+  return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+    DateTimeFormatterBuilder.getLocalizedDateTimePattern(
+      FormatStyle.SHORT,
+      null,
+      IsoChronology.INSTANCE,
+      Locale.getDefault()
+    )
+  } else {
```
Comment (santosh-pingle) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/utilities/MoreLocalDate.kt:49]: do we need else part?
Can you please refer this Java 8+ API desugaring support [Java 8+ API desugaring support ](https://developer.android.com/studio/write/java8-support#library-desugaring)?
I think it is possible to use java.time.* package without android sdk api constraint.

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt (hunk 1 above)
Comment (santosh-pingle) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:137]: Will ```Questionnaire.QuestionnaireItemComponent.dateEntryFormat``` be used to format the selected or entered input from picker, like format the input and show it in input text field

Comment (santosh-pingle) [inline datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:137]: Can we please add this extension to the ```date_picker_questionnaire.json``` and ```date_time_picker_questionnaire.json``` to show this use case?

**Review by santosh-pingle — COMMENTED** [review]
Can you please add some screenshots?

I put some comments, please see whether those can be addressed.

**Comment by PallaviGanorkar** [comment]
> Can you please add some screenshots?
> 
> I put some comments, please see whether those can be addressed.
DateEntryFormatFromJson
DateEntryFormatFromDeviceLocale

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt (hunk 1 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:136]: Added date entry format in date_picker_questionnaire.json only to demonstrate use case 1. If date entry format is available in questionnaire json file.
And not added date entry format in date_time_picker_questionnaire.json to demonstrate use case 2. If date entry format is NOT available in questionnaire json file then read date format from device locale

**Comment by PallaviGanorkar** [comment]
> Can you please add some screenshots?
> 
> I put some comments, please see whether those can be addressed.

Added screen shot

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt (hunk 1 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:136]: As per requirement, It only shows date format as helper text under the date picker

Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:136]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/utilities/MoreLocalDate.kt (hunk 2 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/utilities/MoreLocalDate.kt]: Done

File: datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt (hunk 1 above)
Comment (PallaviGanorkar) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:136]: This is part of validation, We will consider this under input validation ticket.

Comment (santosh-pingle) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/MoreQuestionnaireItemComponents.kt:136]: Ok , can we please create separate issue and link here.

**Comment by santosh-pingle** [comment]
looks ok to me. thanks!

**Review by jingtang10 — CHANGES_REQUESTED** [review]
for the 3 changes in the PR description, 1 and 2 can probably be grouped in this PR. But please raise a different PR for 3. thanks!

**Comment by PallaviGanorkar** [comment]
As per comment crate two separate PR for this change

**Pull request closed without merging**